File: opc_ua_server_pixtend.cpp
// Salzburg Research ForschungsgesmbH
// Armin Niedermueller

// OPC UA Server on PiXtend

#include <iostream>
#include <fstream>
#include <algorithm>
#include <time.h>
#include <string>

#include <thread>
#include <chrono>

#include <opc/ua/node.h>
#include <opc/ua/subscription.h>
#include <opc/ua/server/server.h>



using namespace OpcUa;

class SubClient : public SubscriptionHandler
{
  void DataChange(uint32_t handle, const Node & node, const Variant & val, AttributeId attr) override
  {
    std::cout << "Received DataChange event for Node " << node << std::endl;
  }
};

std::vector<OpcUa::Variant> MyMethod(NodeId context, std::vector<OpcUa::Variant> arguments)
{
  std::cout << "MyMethod called! " << std::endl;
  std::vector<OpcUa::Variant> result;
  result.push_back(Variant(static_cast<uint8_t>(0)));
  return result;
}

void RunServer()
{
  // First setup our server
  auto logger = spdlog::stderr_color_mt("server");
  OpcUa::UaServer server(logger);
  server.SetEndpoint("opc.tcp://localhost:4840/freeopcua/server");
  server.SetServerURI("urn://exampleserver.freeopcua.github.io");
  server.Start();

  // then register our server namespace and get its index in server
  uint32_t idx = server.RegisterNamespace("Namespace - DTZ PiXtend");

  // Create our address space using different methods
  Node objects = server.GetObjectsNode();

  // Add a Node named "Parameters" with the id "1" and in namespace idx (index)
  NodeId nid(1, idx);
  QualifiedName qn("Parameters", idx);
  Node param_object = objects.AddObject(nid, qn);  
    
  // Add a variable and a property with auto-generated nodeid to our custom object
  Node conbelt_state = param_object.AddVariable(idx, "Conveyorbelt - State", Variant(8));
  Node conbelt_distance = param_object.AddVariable(idx, "Conveyorbelt - Distance", Variant(8.8));
  Node conbelt_drive = param_object.AddMethod(idx, "Conveyorbelt - Drive Control", MyMethod);


  // browse root node on server side
  Node root = server.GetRootNode();
  logger->info("Root node is: {}", root);
  logger->info("Children are:");

  for (Node node : root.GetChildren())
    {
      logger->info("    {}", node);
    }


  // Uncomment following to subscribe to datachange events inside server
  /*
  SubClient clt;
  std::unique_ptr<Subscription> sub = server.CreateSubscription(100, clt);
  sub->SubscribeDataChange(conbelt_state);
  */



  // Now write values to address space and send events so clients can have some fun
  conbelt_state.SetValue(Variant("init")); //will change value and trigger datachange event
  conbelt_distance.SetValue(Variant(0.0)); 

  // Create event
  server.EnableEventNotification();
  Event ev(ObjectId::BaseEventType); // you should create your own type
  ev.Severity = 2;
  ev.SourceNode = ObjectId::Server;
  ev.SourceName = "Event from FreeOpcUA";
  ev.Time = DateTime::Current();

  // define filestreams
  std::ifstream statefile("state.log");
  std::ifstream distancefile ("distance.log");
  std::string line;

  logger->info("Ctrl-C to exit");

  for (;;)
    {

	  // read conbelt_state from file and set value inside our object
	  getline(statefile, line);
      conbelt_state.SetValue(Variant(line));		// will change value and trigger datachange event


	  // read conbelt_distance from file and set value inside our object
	  getline(distancefile, line);
	  std::string::size_type sz;			 
	  double distance = std::stod(line, &sz);	
	  conbelt_distance.SetValue(Variant(distance)); // will change value and trigger datachange event


      std::stringstream ss;
	  ss << "This is an conbelt_state change event";
      ev.Message = LocalizedText(ss.str());
      server.TriggerEvent(ev);
      std::this_thread::sleep_for(std::chrono::milliseconds(5000));


    }

  // close filestreams and shut down server
  statefile.close();
  distancefile.close();
  server.Stop();
}

// MAIN

int main(int argc, char ** argv)
{
  try
    {
      RunServer();
    }

  catch (const std::exception & exc)
    {
      std::cout << exc.what() << std::endl;
    }

  return 0;
}

File: opc_ua_server_pixtend.py
#      _____         __        __                               ____                                        __
#     / ___/ ____ _ / /____   / /_   __  __ _____ ____ _       / __ \ ___   _____ ___   ____ _ _____ _____ / /_
#     \__ \ / __ `// //_  /  / __ \ / / / // ___// __ `/      / /_/ // _ \ / ___// _ \ / __ `// ___// ___// __ \
#    ___/ // /_/ // /  / /_ / /_/ // /_/ // /   / /_/ /      / _, _//  __/(__  )/  __// /_/ // /   / /__ / / / /
#   /____/ \__,_//_/  /___//_.___/ \__,_//_/    \__, /      /_/ |_| \___//____/ \___/ \__,_//_/    \___//_/ /_/
#                                              /____/
# Salzburg Research ForschungsgesmbH
# Armin Niedermueller

# OPC UA Server on PiXtend
# The purpose of this OPCUA server is to provide methods to control the conveyorbelt (stepper motor) and read the state
# of the conveyor belt
# the hardware is PiXtend - Raspberry Pi SPS

import sys
sys.path.insert(0, "..")
import time
import logging
import socket

from ConveyorBeltDummy import ConveyorBeltDummy as ConveyorBeltX
#from ConveyorBeltX import ConveyorBeltX
from opcua import ua, uamethod, Server
import datetime
import time
import threading
import traceback

# setup conveyor belt
conbelt = ConveyorBeltX()

def move_belt_core(direction, distance):
    if direction == "right":
        conbelt.move_right_for(distance)
    elif direction == "left":
        conbelt.move_left_for(distance)
    return True

# Method to control the conveyor belt - direction is "right" or "left", distance is float in meters.
@uamethod
def move_belt(parent, direction, distance):
    move_thread = threading.Thread(name='move_belt_thread', target = move_belt_core, args = (direction, distance,))
    move_thread.daemon = True
    move_thread.start()
    return True

# method for alarm light control - Value busy True means the light should show red, e.g when the robot is running
@uamethod
def switch_light(parent, busy):
    conbelt.busy_light(busy)
    return True

@uamethod
def reset_totaldist(parent):
    conbelt.reset_totaldistance(0)
    conbelt.total_distance = 0
    conbelt.distance = 0

if __name__ == "__main__":
    logger = logging.getLogger(__name__)
    logging.basicConfig(format='%(asctime)s,%(msecs)d %(levelname)-8s %(name)s [%(filename)s:%(lineno)d] %(message)s', level=logging.INFO, datefmt='%Y-%m-%d %H:%M:%S')
    logging.getLogger("opcua.server").setLevel(logging.ERROR)
    logger.info("Starting OPC-UA Server on host: {}"
                .format(socket.gethostname()))
    
    # setup our server
    server = Server()
    url = "opc.tcp://0.0.0.0:4840/freeopcua/server"
    server.set_endpoint(url)

    # setup our own namespace
    server.register_namespace("ns=2")


    # get Objects node, this is where we should put our nodes
    objects = server.get_objects_node()

    # Add a parameter object to the address space
    conveyorbelt_object = objects.add_object("ns=2; i=1", "ConveyorBelt")

    # Parameters - Addresspsace, Name, Initial Value
    server_time = conveyorbelt_object.add_variable("ns=2; i=2", "ServerTime", 0)
    mover = conveyorbelt_object.add_method("ns=2; i=3", "MoveBelt", move_belt, [ua.VariantType.String, ua.VariantType.Float], [ua.VariantType.Boolean])
    resetTotalDistance = conveyorbelt_object.add_method("ns=2; i=14", "ResetTotalDistance", reset_totaldist)
    busy_light = conveyorbelt_object.add_method("ns=2; i=7", "SwitchBusyLight", switch_light, [ua.VariantType.Boolean] , [ua.VariantType.Boolean])
    conbelt_state = conveyorbelt_object.add_variable("ns=2; i=10", "ConBeltState", conbelt.state)
    conbelt_dist = conveyorbelt_object.add_variable("ns=2; i=11", "ConBeltDist", conbelt.distance)
    conbelt_moving = conveyorbelt_object.add_variable("ns=2; i=12", "ConBeltMoving", False)
    conbelt_totaldist = conveyorbelt_object.add_variable("ns=2; i=13", "ConBeltTotalDist", conbelt.total_distance)
    conbelt_service_order_notification = conveyorbelt_object.add_variable("ns=2; i=15", "ConBeltServiceOrderRequest", conbelt.maintenance_required)
    conbelt_failure = conveyorbelt_object.add_variable("ns=2; i=16", "ConBeltFailure", conbelt.state == "fail")

    # Set parameters writable by clients
    server_time.set_writable()


    # Start the server
    server.start()

    logger.info("OPCUA - Pixtend - Server started at {}".format(url))

    try:
        while True:
            TIME = datetime.datetime.now()  # current time
            if conbelt.state != "init" and conbelt.state != "stop" and conbelt.state != "halt" and conbelt.state != "fail":
                conbelt_moving.set_value(True)
            else:
                conbelt_moving.set_value(False)

            if conbelt.state == "fail":
                conbelt_failure.set_value(True)
            else:
                conbelt_failure.set_value(False)

            # set the random values inside the node
            logger.debug("Belt-State: " + str(conbelt.state) + "   Belt-Distance: " + str(conbelt.distance)+ "   Belt-Total Distance: " + str(conbelt.total_distance) + "   Server-Time: " + str(server_time.get_value()))
            server_time.set_value(TIME)
            conbelt_state.set_value(conbelt.state)
            conbelt_dist.set_value(conbelt.distance)
            conbelt_totaldist.set_value(conbelt.total_distance)
            conbelt_service_order_notification.set_value(conbelt.maintenance_required)

            # sleep 2 seconds
            time.sleep(2)
    except KeyboardInterrupt:
        logger.info("\nCtrl-C pressed. OPCUA - Pixtend - Server stopped at {}".format(url))
    except Exception as ex:
        traceback.print_exc()
        logger.error(ex)
    finally:
        #close connection, remove subcsriptions, etc
        server.stop()
        conbelt = None
        sys.exit(0)
